### ISSUE TITLE
Fix knot-tying via an explicit recursive combinator

### ISSUE THREAD

**Issue by tel** [tel/ocaml-fstreams#6]
Right now the internal implementations of streams are exposed so that people can do "stupid stream tricks" like

``` ocaml
let rec fibs = 
  lazy { head = 0
       ; tail =
           lazy
             { head = 1
             ; tail = map2 (+) fibs (tail fibs)
             }
       }
```

This can't be made to work using the API like

``` ocaml
let rec fibs = cons 0 (cons 1 (map2 (+) fibs (tail fibs)))
```

since `let rec` won't let you refer to a value recursively, only a function. If we change `fibs` to a function then it'll loose all sharing and be useless.

What might be useful to work around this is a combinator like

``` ocaml
val mu : ('a stream -> 'a stream) -> 'a stream
```

which gives

``` ocaml
let fibs = mu (fun s -> cons 0 (cons 1 (map2 (+) s (tail s))))
```

In other words, we'll do the recursion hacking internally to the API.

**Comment by tel** [comment]
Since OCaml has physical equality, as ill-advised as it might be, we can actually observe the difference between "fast" and "large" structures. This makes for some great tests!

``` ocaml
let cyc = mu (fun s -> cons 1 (cons 2 (cons 3 s))) in
let (h1, t1) = uncons cyc in
let (h2, t2) = uncons cyc in
let (h3, t3) = uncons cyc in
cyc == t3
```

**Comment by tel** [comment]
I _think_ I was able to achieve this

``` ocaml
    let mu f_ =
      let rec x = lazy begin
        Lz.from_fun begin fun () ->
          Lz.force (f_ (Lazy.Memoized.force x))
        end
      end
      in Lazy.Memoized.force x
```

and it works for both instantiations of `Lz` without exposing the lazy mutation.
